pkg/logs/input/file/tailer.go

```diff
@@ -43,9 +45,11 @@ type Tailer struct {
 
 	outputChan  chan *message.Message
 	decoder     *decoder.Decoder
-	source      *config.LogSource
 	tagProvider tag.Provider
 
+	sourceMu sync.Mutex
```

Comment:
I agree with your comment that maybe there could be a better way to do this structurally (I am also thinking about this). 

But one small possible improvement could be to use `sync.RWMutex` in this case since writes will be very uncommon compared to reads/access (especially in the `readForever` loop).